crates/walrus-service/src/common/telemetry.rs

```diff
@@ -321,12 +327,49 @@ fn http_version(request: &axum::extract::Request) -> &'static str {
     }
 }
 
+#[derive(Clone, Debug)]
+pub(crate) struct MetricsMiddlewareState {
+    inner: Arc<MetricsMiddlewareStateInner>,
+}
+
+impl MetricsMiddlewareState {
+    pub fn new(registry: &Registry) -> Self {
+        Self {
+            inner: Arc::new(MetricsMiddlewareStateInner {
+                metrics: HttpServerMetrics::new(registry),
+                task_monitors: TaskMonitorFamily::new(registry.clone()),
+            }),
+        }
+    }
+
+    /// Returns the task monitor for the route and request, where `http_route` should be a
+    /// known route or `UNMATCHED_ROUTE`.
+    fn task_monitor(&self, method: Method, http_route: &str) -> TaskMonitor {
+        self.inner
+            .task_monitors
+            .get_or_insert_with_task_name(&(method.clone(), http_route.to_owned()), || {
```

Comment:
Question: what's the overhead of a monitor? Looks like we are creating one monitor per route, and the number of metrics tracked seems a lot.